dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(..))
+import qualified Dhall.Map as DM (keys)
+import Dhall.Parser
+import Data.Text (Text)
+import Control.Monad
+import Control.Exception
+import qualified Data.Text as T
+import qualified Data.Text.IO as TIO
+import qualified Data.Map as M
+import Data.List (isSuffixOf)
+import System.Directory (doesDirectoryExist, getDirectoryContents,
+#if MIN_VERSION_directory(1,3,0)
+                         pathIsSymbolicLink)
+#else
+                         isSymbolicLink)
+#endif
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+data FilePos = FP Int Int deriving (Eq, Ord)
+newtype Tags = Tags (M.Map FilePath (M.Map FilePos Tag))
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith M.union ts1 ts2)
+data Tag = Tag Text Text
+
+{-| Generate ETags for Dhall expressions
+-}
+generate :: Input -> [Text] -> Bool -> IO Text
+generate StandardInput _ _ = showTags . tags "" <$> TIO.getContents
+generate (File p) sxs followSyms = do
+    isD <- doesDirectoryExist p
+    files <- dirToFiles followSyms (map T.unpack (if isD then sxs else [""])) p
+    showTags <$> foldM (\ts f -> do
+                                   t <- tags f <$> handle (\(SomeException _) -> return "")
+                                                          (TIO.readFile f)
+                                   return (ts <> t)) (Tags M.empty) files
+
+tags :: FilePath -> Text -> Tags
+tags f t = Tags (M.singleton f
+                 (M.union initialMap (parse t)))
+    where initialMap = M.fromList [ (FP 1 0, Tag "" (T.pack . takeFileName $ f)) -- vi
+                                  , (FP 1 1, Tag "" ("/" <> (T.pack . takeFileName) f)) -- emacs
```

Comment:
What do the comments `vi` and `emacs` mean?

Reply:
Added more clear description.